### PR TITLE
SelectArrayGently plugin

### PR DESCRIPTION
**Pull request by Antikon**
### What are you changing/introducing
A new ngrest plugin which is similar to SelectArray plugin.

### What is the reason for changing/introducing
The SelectArray plugin has bad behavior, which leads to a change in the model attribute (see #439). This behavior can be disabled by assignAfterFind option (#444). Howewer, this option disables autoconversion of data in a list view also.

SelectArrayGently plugin has the advantages of original SelectArray plugin and has not its disadvantages.

### QA

| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | no
| Tests pass?   | yes
| Fixed issues  | #490

### CODE REVIEW

File: src/ngrest/plugins/SelectArrayGently.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+ * @author Anton Ikonnikov <antikon2@yandex.ru>
+ * @since 4.1.0
+ */
+class SelectArrayGently extends Select
+{
+    use LazyDataLoadTrait;
+    
+    private $_data;
+
+
```
Comment (nadar) [inline src/ngrest/plugins/SelectArrayGently.php:48]: please remove the empty break and add @inheritdoc

File: src/ngrest/plugins/SelectArrayGently.php [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+     * @inheritdoc
+     */
+    public function renderList($id, $ngModel)
+    {
+        if ($this->scheduling && $this->renderContext->canUpdate()) {
+            return $this->createSchedulerListTag($ngModel, $this->getData(), 'item');
+        }
+
+        $options = Angular::optionsFilter([
+                                              'options' => $this->getServiceName('selectdata'),
```
Comment (nadar) [inline src/ngrest/plugins/SelectArrayGently.php:61]: some empty whitespaces here

**Review by nadar — CHANGES_REQUESTED** [review]
Hey @Antikon - thats a great change, thank you very much - even though the name "gently" is uncommon, i like it :+1: 

i just have added some phpdoc request changes, and please add a changelog. thanks!

File: src/ngrest/plugins/SelectArrayGently.php (hunk 1 above)
Comment (Antikon) [reply src/ngrest/plugins/SelectArrayGently.php:50]: added some comments (@inheritdoc refers to nowhere for $_data)

File: src/ngrest/plugins/SelectArrayGently.php (hunk 2 above)
Comment (Antikon) [reply src/ngrest/plugins/SelectArrayGently.php]: done

**Review by nadar — CHANGES_REQUESTED** [review]
One thing, you write  `...override the values from the database.` but i think you mean that value of that attributes does not recieve "the value" and recieves the "key" instead? :flushed: 

Or are you reffering to your example code `new Query()->all()`? Sorry for the delay

**Comment by Antikon** [comment]
Sorry, I didn't get exactly what you mean.

Initial 'SelectArray' plugin changes the model's attributes values which came from DB using `onAfterListFind` method of abstract `Select` class. In the 'SelectArrayGently' plugin this behavior is disabled, so the values of the attributes remain unchanged. By `...override the values from the database` I meant exactly that behavior. 
Maybe the word _default_ should be added as it says in the Guide.

**Comment by nadar** [comment]
Ok, i see. I think my description in `$assignAfterFind` was chosen bad. Because the attribute will be overwritte with the `label` defined in $data array, which is not directly "the value from the database". https://github.com/luyadev/luya-module-admin/blob/master/src/ngrest/plugins/Select.php#L117

:+1: 

Could we then not just make a new class like

```php
class SelectArrayGently from SelectArray
{
   public $assignAfterFind = false;
}
```

Is there any difference to your code? Is the new Angular Directive Required?

**Comment by Antikon** [comment]
Ок, let me explain once more.

Let's assume we have 'Country' and 'City' models.
In a last one we have a relation

'''
public function getCountry()
    {
        return $this->hasOne(ConfCountries::class, ['id' => 'country_id']);
    }
'''
We also have in 'City' model
'''
public function ngRestAttributeTypes()
{
    'country_id' => [
        'selectArray',
        'data'       => ConfCountries::getCountriesArray(), // This is just example of array, e.g. [1 => 'France', '2' => 'Germany', etc]
    ],
}
'''
'country_id' is listed ngRestScopes() of 'City' model.

'SelectArray' plugin renders a dropdown in 'create' or 'update' context. And shows country name in 'list' context. This is what we expect from it. But at the same time in 'list' context it changes the values of 'country_id' attribute of 'City' model (Was 1, now France). This is exactly what I call 'override the values from the database'.

This behavior can be disabled by '$assignAfterFind = false'. In that case a dropdown still exist in 'create' or 'update' context, but in 'list' context the country id not the country name will be shown. In order to avoid this, I do the conversion in the 'SelectArrayGently' plugin on the client side by new AngularJs directive.

Thus, the behavior of the 'SelectArrayGently' plugin is the same as 'SelectArray' plugin, but the initial value of the attribute ('country_id' in our example) remains unchanged.

**Comment by nadar** [comment]
> In order to avoid this, I do the conversion in the SelectArrayGently plugin on the client side by new AngularJs directive.

:+1:

**Review by nadar — APPROVED** [review]

**Pull request merged**